modules/chemical_reactions/src/userobjects/ThermochimicaData.C

```diff
@@ -139,17 +148,22 @@ ThermochimicaNodalData::ThermochimicaNodalData(const InputParameters & parameter
   }
 }
 
+template <bool is_nodal>
 void
-ThermochimicaNodalData::initialize()
+ThermochimicaDataBase<is_nodal>::initialize()
 {
 }
 
+template <bool is_nodal>
 void
-ThermochimicaNodalData::execute()
+ThermochimicaDataBase<is_nodal>::execute()
 {
 #ifdef THERMOCHIMICA_ENABLED
-  auto temperature = _temperature[_qp];
-  auto pressure = _pressure[_qp];
+  // either one DOF at a node or (currently) one DOF for constant monomial FV!
```

Comment:
Is this lower order restriction checked somewhere else?

Reply:
We'd need to make sure that we have one DOF per node and one DOF per element. Can I add a simple MooseArray::size check here?

Reply:
Yeah, probably the easiest choice. I think that a real error (not an assertion) is more appropriate, but it's your call

Reply:
Nah. `_temperature.size() == 216`...

Reply:
I can use `var->numberOfDofs()` but only in `execute()`. I should be able to figure out if the coupled var  is suitable in the constructor, as we cannot have the same variable have different families or orders on different blocks.

Reply:
yeah just use the variable order in the constructor?

Reply:
And family I suppose? And element order? Second order Lagrange is fine. Any Lagrange var has one dof per node, right? What about the other families?

Reply:
Ok, this is untestable because these UOs can only be built using the `ChemicalCompositionAction`, which automatically sets up the correct variables. I added a comment to that effect.